cylc/flow/suite_srv_files_mgr.py

```diff
@@ -33,642 +34,692 @@
-* cylc stop --kill %(suite)s       # kill active tasks and wait
-* cylc stop --now %(suite)s        # don't wait for active tasks
-* cylc stop --now --now %(suite)s  # don't wait
-* ssh -n "%(host)s" kill %(pid)s   # final brute force!
-"""
-            ) % {
-                "host": old_host,
-                "port": old_port,
-                "pid": old_pid_str,
-                "fname": fname,
-                "suite": reg,
-            }
-        )
+        DIRNAME = '.service'
+        """The name of this directory."""
 
-    def dump_contact_file(self, reg, data):
-        """Create contact file. Data should be a key=value dict."""
-        # Note:
-        # 1st fsync for writing the content of the contact file to disk.
-        # 2nd fsync for writing the file metadata of the contact file to disk.
-        # The double fsync logic ensures that if the contact file is written to
-        # a shared file system e.g. via NFS, it will be immediately visible
-        # from by a process on other hosts after the current process returns.
-        with open(self.get_contact_file(reg), "wb") as handle:
-            for key, value in sorted(data.items()):
-                handle.write(("%s=%s\n" % (key, value)).encode())
-            os.fsync(handle.fileno())
-        dir_fileno = os.open(self.get_suite_srv_dir(reg), os.O_DIRECTORY)
-        os.fsync(dir_fileno)
-        os.close(dir_fileno)
-
-    def get_contact_file(self, reg):
-        """Return name of contact file."""
-        return os.path.join(
-            self.get_suite_srv_dir(reg), self.FILE_BASE_CONTACT)
+        CONTACT = 'contact'
+        """Contains settings for the running suite.
 
-    def get_auth_item(self, item, reg, owner=None, host=None, content=False):
-        """Locate/load passphrase, SSL private key, SSL certificate, etc.
+        For details of the fields see ``ContactFileFields``.
+        """
 
-        Return file name, or content of file if content=True is set.
-        Files are searched from these locations in order:
+        CONTACT2 = 'contact2'
+        """Same as ``CONTACT``, installed on remote platforms."""
 
-        1/ For running task jobs, service directory under:
-           a/ $CYLC_SUITE_RUN_DIR for remote jobs.
-           b/ $CYLC_SUITE_RUN_DIR_ON_SUITE_HOST for local jobs or remote jobs
-              with SSH messaging.
+        PASSPHRASE = 'passphrase'
+        """The suite authentication token."""
 
-        2/ (Passphrases only) From memory cache, for remote suite passphrases.
-           Don't use if content=False.
+        SOURCE = 'source'
+        """Symlink to the suite definition (suite dir)."""
 
-        3/ For suite on local user@host. The suite service directory.
 
-        4/ Location under $HOME/.cylc/ for remote suite control from accounts
-           that do not actually need the suite definition directory to be
-           installed:
-           $HOME/.cylc/auth/SUITE_OWNER@SUITE_HOST/SUITE_NAME/
+class ContactFileFields:
```

Comment:
Couldn't you just use a dictionary instead of a class here? (and for the other similar classes).

Reply:
(Are the attribute doc strings are the reason? - even though they're not a real thing as far as Python is concerned)

Reply:
I had a similar comment sitting on my pending review but forgot to submit it. Looks like we ended up with similar questions on the constants :+1: still need to "view" a few more files according to GitHub's UI, but I think they are most cosmetics due to changes in imported names.

Reply:
> Are the attribute doc strings are the reason

Yep, Sphinx can auto document the attributes. I think perhaps there might be a nicer way to do this in #3345.

Reply:
It seems a little odd to let Sphinx dictate how we use Python, to the extent of using classes where dictionaries would be more straightforward and attribute docstrings that are proper docstrings (really just comments, effectively) to Python.

However, in context the benefits are worth it I guess :thinking:   Probably worth a comment in these classes though, to explain the reasoning to future us (or others).

Reply:
> It seems a little odd to let Sphinx dictate how we use Python ...

I guess modern expectation of the ecosystem of a software project has outpaced the Python language itself. Support for smart comments would have been nice in this case. (This may not be so far off, as I would never have thought that Python would support type hint, for example.)

Reply:
> using classes where dictionaries would be more straightforward

Would a dictionary actually be more straight forward though?

```python
# dictionary solution
SUITE_FILES = {
    'suite rc': 'suite.rc',  # the suite definition file
    # ...
}

# formal enum
class SuiteFiles(Enum):
    SuiteRC = 'suite.rc'
    """The suite definition file."""

# informal enum
class SuiteFiles:
     SuiteRC = 'suite.rc'
    """The suite definition file."""
```

What's nicer:

```python
# pulling values out of a dictionary,
SUITE_FILES['suite.rc']

# or referencing items on an object?
SuiteFiles.SuiteRC
```

The dict solution can be dangerous:

```python
SUITE_FILES['suite.rc']  # error - the key is "suite rc"

# to avoid this you end up with top level constants floating in the module
from suite_files import SUITE_RC
```

Classes aren't necessarily stateful objects, they can be static data structures too. In this case the class is just an enumeration.

Enumeration bonuses:

* Formal declaration of constants.
* Code checkers can detect if you are referencing keys which don't exist.
* Documentation both for developers and (via auto-generation) for users.
* Additional functionality e.g. parsing, validation, operations on enumerations.

Besides the enum implementation is basically just syntactic sugar, it's just a dictionary underneath:

```python
>>> SuiteFiles.__dict__
mappingproxy({'__module__': '__main__', 'SuiteRC': 'suite.rc', ... >, '__doc__': None})
```

Here is an excerpt from my task state tidy branch, (sadly abandoned due to protobuff limitations):

```python
class TaskStatus(Enum):
    """Task status names and meanings.

    Tasks statuses are ordered according to task runtime progression.
    """

    RUNAHEAD = "runahead"
    """Held back from dependency matching, in the runahead pool."""

    WAITING = "waiting"
    """Held back from job submission due to un-met prerequisites."""

    QUEUED = "queued"
    """Prerequisites met, but held back in a limited internal queue."""

    # ...

    def __gt__(self, other):
        order = list(self.__class__.__members__.values())
        return order.index(self) > order.index(other)

    def __lt__(self, other):
        order = list(self.__class__.__members__.values())
        return order.index(self) < order.index(other)

    # ...
```

```python
>>> for task_status in TaskStatus:
...     # no need for a TASK_STATUSES_ORDERED list
...     pass
>>> TaskStatus.QUEUED > TaskStatus.RUNAHEAD
True
```

Reply:
Alright, I'm pretty much convinced.  My "more straightforward" point was more a matter of principle - surely the whole point of a "class" is to act as a template for objects.  Whereas a dictionary is a just a data structure. (The fact that classes are implemented with dictionaries - plus extra machinery btw - is just an implementation detail).